Add "Extended" cheat menu and new "All Doors Unlocked" cheat

johnmcollier opened the pull request:
Hi! 

This PR adds a basic "Unlock all doors" cheat to the game,

I initially prepared this patch a year ago just for my own fun while I played Perfect Dark on PC, as I was always interested in the various doors in Perfect Dark that were locked and had no key. I opted to create a new cheat inspired by the "Dual Weapons" cheat that was added to the PC port.

The way the cheat works is fairly straightforward, when it's enabled, whenever a new level is loaded, the game ignores all locked door checks in that level. It does this, by adding a check for the cheat's enablement to `doorIsUnlocked()` in `src/propobj.c`. When the cheat is enabled, this function always returns true, otherwise the function's existing logic is used.

Unlike the dual weapons cheat, I struggled to find a good way to add the cheat text to the game. I tried to use the existing "Door unlocked" string present in the game, but it didn't work as it didn't seem to be loaded when accessing the cheat menu. Instead, I opted to just hardcode the cheat text as a constant in cheats.c (see https://github.com/johnmcollier/perfect_dark/blob/fd737351a7f97249d71cc8668858ab9430c87e47/src/game/cheats.c#L26). I'm still relatively new to the source code here, so I'm sure there's other/better ways to handle this, let me know.

For an example of the cheat in action, see here: https://streamable.com/1ifx8m

Anyways, let me know if this is an okay patch to include!

Edit:

Based on PR feedback, I've also included a new "Extras" cheat menu to put the cheat under, and any potential new cheats added to the PC port (like dual weapons).

johnmcollier committed:
Add "All Doors Unlocked" cheat

Signed-off-by: John Collier <6880023+johnmcollier@users.noreply.github.com>

Atari-Dude commented:
imo a new separate menu should be created for things which are cool to experiment with, but can easily break the game (and/or are experimental/debug/development features), rather than being included with the typical cheats menu. So that there's a clear distinction for the average player that some cheats/options are not as battle tested as the original cheats, whilst still allowing their inclusion.

Such things could include the (Two-Handed) cheat since its implementation is pretty janky, even if it's cool. All Doors Unlocked, because some doors you'll just fall into a void after you step out (i.e. in the dD lobby, and Grid). The ability to deploy multiple Laptop sentries from #446. Unlimited Psychosis Gun ammo like in #301. Moon jump, or free cam. The hot-reload feature for external textures in current test builds of Raf's forthcoming release (which is a useful feature for updating textures in real time)

Stuff like that.

johnmcollier committed:
Create new "Extended Cheats" window

Signed-off-by: John Collier <6880023+johnmcollier@users.noreply.github.com>

johnmcollier committed:
Remove NTSC restriction on Doors Unlocked cheat

Since the Extended menu itself isn't localized, I think we're fine to remove NTSC-only restriction from this cheat.

Signed-off-by: John Collier <6880023+johnmcollier@users.noreply.github.com>

BiatuAutMiahn commented:
Why not name it 'Mods' instead of cheats? small rant: 'Cheats' in PD were always weird..thowing "Cheat" all over the menu made it feel like strings are attached to the cheats enabled, where most games just activated a particular feature and never squawked about it.

johnmcollier committed:
Properly support the cheat on PAL and JPN versions

Previously we were only handling the All Doors Cheat name literal in just NTSC 1.0 and final. This updates things so that all versions of the game handle the literal.

Signed-off-by: John Collier <6880023+johnmcollier@users.noreply.github.com>

johnmcollier commented:
> Why not name it 'Mods' instead of cheats? small rant: 'Cheats' in PD were always weird..thowing "Cheat" all over the menu made it feel like strings are attached to the cheats enabled, where most games just activated a particular feature and never squawked about it.

Since this (and the dual weapons cheat) is the using the existing cheat functionality in the game's code, I figured it was best to keep the menu labeled as a Cheat menu, at least for now

johnmcollier committed:
Allow custom cheats to use string literals

Adds a new lookup table that contains string literals for each of the custom cheats in the game. `cheatGetName` and `cheatGetNameIfUnlocked` are also updated accordingly to fetch the custom string if called on a custom cheat.

Signed-off-by: John Collier <6880023+johnmcollier@users.noreply.github.com>
Signed-off-by: John Collier <jcollier@redhat.com>
